datadog/fwprovider/resource_datadog_integration_gcp_sts.go

```diff
@@ -0,0 +1,428 @@
+		enableCSPM = plan.EnableCspm.ValueBool()
+	}
+
+	saInfo := datadogV2.GCPServiceAccountCreateRequestData{
+		Data: &datadogV2.GCPServiceAccountData{
+			Attributes: &datadogV2.GCPServiceAccountAttributes{
+				ClientEmail:   stringToPointer(plan.ServiceAccountEmail.ValueString()),
+				Automute:      boolToPointer(enableAutomute),
+				IsCspmEnabled: boolToPointer(enableCSPM),
+				HostFilters:   hostFilters,
+			},
+			Type: datadogV2.GCPSERVICEACCOUNTTYPE_GCP_SERVICE_ACCOUNT.Ptr(),
+		},
+	}
+
+	createResponse, _, err := r.GcpApi.CreateGCPSTSAccount(r.Auth, saInfo)
+	if err != nil {
+		resp.Diagnostics.AddError("Error creating STS service account",
+			"Error creating an entry within Datadog for your STS enabled service account: "+err.Error())
+		return
+	}
+	createdServiceAccountInfo := createResponse.GetData()
+
+	// Set the "computed" values.
+	plan.ID = types.StringValue(createdServiceAccountInfo.GetId())
+	plan.DelegateEmail = types.StringValue(delegateAttributes.GetDelegateAccountEmail())
+
+	diags = resp.State.Set(ctx, plan)
+	resp.Diagnostics.Append(diags...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+}
+
+// Read re-sets the Terraform state using the latest "pulled" data.
+func (r *datadogIntegrationGCPSTSResource) Read(ctx context.Context, req resource.ReadRequest, resp *resource.ReadResponse) {
+	var state datadogIntegrationGCPSTSResourceModel
+	diags := req.State.Get(ctx, &state)
+	resp.Diagnostics.Append(diags...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	delegateResponse, _, err := r.GcpApi.GetGCPSTSDelegate(r.Auth)
+	if err != nil {
+		resp.Diagnostics.Append(utils.FrameworkErrorDiag(err, "error retrieving STS delegate"))
+		return
+	}
+
+	delegateEmail, err := extractDelegateAccountEmail(delegateResponse)
+	if err != nil {
+		resp.Diagnostics.Append(utils.FrameworkErrorDiag(err, "error getting delegate account email"))
+		return
+	}
+	state.DelegateEmail = delegateEmail
+
+	stsEnabledAccounts, _, err := r.GcpApi.ListGCPSTSAccounts(r.Auth)
+	if err != nil {
+		resp.Diagnostics.AddError("Error retrieving STS service accounts",
+			"Error listing GCP STS Accounts: "+err.Error())
+		return
+	}
+
+	foundAccount, err := findServiceAccountByUniqueID(stsEnabledAccounts, state.ID.ValueString())
+	if err != nil {
+		resp.Diagnostics.AddError("Error finding your service account", err.Error())
+		return
+	}
+
+	outputListValue, hostFilterCount := getHostFilters(foundAccount)
+
+	// The section below handles optional fields in Terraform.
+	// If an optional field is not specified, then Terraform state stores a nil value.
+	// However, the Datadog GCP API always responds with a value for these optional fields (an empty list, a false boolean, etc).
+	// If these optional fields aren't specified within Terraform Resources, then these fields should always remain nil.
+	if state.HostFilters.IsNull() {
+		if hostFilterCount > 0 {
+			state.HostFilters = outputListValue
+		}
+	} else {
+		state.HostFilters = outputListValue
```

Comment:
Will outputListValue ever by an empty list? Will Terraform convert that to Null?

Reply:
Yes, outputListValue will be an empty list if the service account created in Datadog's backend was created without specifying host filters, or if the service account was created with host filters explicitly set to an empty list. 

Terraform won't convert the empty list to null. This isn't an issue, however a user will always see a change in a Terraform plan (from nil -> empty list) if they didn't choose to use the optional host filters field. So this logic section prevents this from showing up